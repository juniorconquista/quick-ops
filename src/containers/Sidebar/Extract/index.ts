import ExtractItem from './extract';

export default ExtractItem;
