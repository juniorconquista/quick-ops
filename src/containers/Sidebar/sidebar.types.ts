import { LayoutProps } from 'styled-system';

export type SidebarStyledProps = LayoutProps & {
    styling?: 'base';
};

export interface SidebarProps {
    isOpen: boolean;
}

export type account = {
    currency: {
        symbol: string;
    };
    balance: number;
    returned: number;
    received: number;
    declared: number;
    approved: number;
    updatedOn: number;
};

export interface SidebarResponse {
    content: account[];
}
