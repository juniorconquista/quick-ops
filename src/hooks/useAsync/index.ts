import useAsync from './useAsync';
export * from './useAsync.utils';

export default useAsync;
