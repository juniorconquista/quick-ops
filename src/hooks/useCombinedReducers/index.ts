import useCombinedReducers from './useCombinedReducers';

export default useCombinedReducers;
