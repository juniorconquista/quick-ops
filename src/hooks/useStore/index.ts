import useStore from './useStore';
export * from './useStore';

export default useStore;
