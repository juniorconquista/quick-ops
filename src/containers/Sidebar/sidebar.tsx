import React, { memo, useEffect } from 'react';
import classnames from 'classnames';
import { Box, H1, H2 } from 'quarks-ui';
import styled from 'styled-components';
import { layout, space, variant, VariantArgs } from 'styled-system';

import { ReactComponent as Arrow } from '../../assets/icons/up-arrow.svg';
import Status from '../../components/Status';
import Loading from '../../components/Loading';
import useWindowSize from '../../hooks/useWindowSize';
import useAsync, { headers } from '../../hooks/useAsync';

import {
    SidebarProps,
    SidebarResponse,
    SidebarStyledProps,
} from './sidebar.types';
import Svg from '../../components/Svg';
import Extract from './Extract';

/* TODO This component was created in the local repository as it has not yet been 
implemented in lib Quarks UI, however in the future it can be changed to be used directly from lib. */

const variants = variant({
    prop: 'styling',
    variants: {
        base: {
            height: '100%',
            '.content': {
                display: 'flex',
                flexDirection: 'column',
                justifyContent: 'flex-start',
                alignItems: 'center',
                minHeight: 'calc(100vh - 48px)',
                height: '100%',
                bg: 'white.default',
                transition: 'right .25s ease',
                position: 'relative',
                '&:focus': {
                    outline: 'none',
                },
            },
            '.mobile': {
                top: 48,
                right: '-350px',
                position: 'fixed',
                width: '350px',
                overflow: 'auto',
                minHeight: 'inherit',
                paddingBottom: '30px',
            },
            '.mobile.open': {
                right: '0',
            },
        },
    },
} as VariantArgs);

const Style: React.FC<SidebarStyledProps> = styled.div<SidebarStyledProps>`
    ${variants}
    ${space}
    ${layout}
`;

const Sidebar: React.FC<SidebarProps> = (props) => {
    const getSidebar = (): Promise<SidebarResponse> =>
        fetch(`${process.env.REACT_APP_API_URL}/sidebar`, {
            headers,
        }).then((response) => response.json());

    const size = useWindowSize();
    const { execute, status, value } = useAsync<SidebarResponse>(
        getSidebar,
        false,
    );

    useEffect(() => {
        execute();
    }, []); // eslint-disable-line react-hooks/exhaustive-deps

    return (
        <Style styling="base">
            <div
                className={classnames({
                    content: true,
                    mobile: size.isMobile,
                    open: size.isMobile && props.isOpen,
                })}
            >
                <Box styling="column" p="40px 0px" width="100%">
                    <Status />
                    {value?.content?.map((account, index) => (
                        <Box
                            key={`${account.updatedOn} - ${index}`}
                            styling="column"
                            mt="30px"
                            width="100%"
                        >
                            <H2 styling="base" fontSize="14px">
                                VALOR A RECEBER
                            </H2>
                            <H1
                                styling="base"
                                color="text.contrast"
                                fontWeight="bold"
                                fontSize="24px"
                                pt="10px"
                            >
                                {`${account.currency.symbol} ${account.balance}`}
                            </H1>
                            <Box
                                styling="row"
                                height="1px"
                                width="150px"
                                border="solid 1px #dfe2e6"
                                m="24px 0px"
                                opacity=".6"
                            />
                            <Box styling="row">
                                <Box styling="row">
                                    <Svg styling="base" mr="15px">
                                        <Arrow height="20px" width="20px" />
                                    </Svg>
                                    <Box
                                        styling="column"
                                        alignItems="flex-start"
                                    >
                                        <H2
                                            styling="base"
                                            fontSize="12px"
                                            mb="5px"
                                        >
                                            Gastou
                                        </H2>
                                        <H1
                                            styling="base"
                                            color="text.contrast"
                                            fontWeight="normal"
                                            fontSize="14px"
                                        >
                                            {`${account.currency.symbol} ${account.returned}`}
                                        </H1>
                                    </Box>
                                    <Box
                                        styling="row"
                                        height="30px"
                                        width="1px"
                                        border="solid 1px #dfe2e6"
                                        m="0px 15px"
                                        opacity=".6"
                                    />
                                    <Svg styling="base" mr="15px">
                                        <Arrow height="20px" width="20px" />
                                    </Svg>
                                    <Box
                                        styling="column"
                                        alignItems="flex-start"
                                    >
                                        <H2
                                            styling="base"
                                            fontSize="12px"
                                            mb="5px"
                                        >
                                            Recebeu
                                        </H2>
                                        <H1
                                            styling="base"
                                            color="text.contrast"
                                            fontWeight="normal"
                                            fontSize="14px"
                                        >
                                            {`${account.currency.symbol} ${account.received}`}
                                        </H1>
                                    </Box>
                                </Box>
                            </Box>
                            <Extract
                                currencySymbol={account.currency.symbol}
                                declared={account.declared}
                                approved={account.approved}
                            />
                        </Box>
                    ))}
                </Box>
                {status === 'pending' && <Loading />}
            </div>
        </Style>
    );
};

export default memo(Sidebar);
