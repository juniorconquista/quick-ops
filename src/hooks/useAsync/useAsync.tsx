import { useState, useEffect, useCallback } from 'react';
import { Status } from './useAsync.types';

const useAsync = <T, E = string>(
    asyncFunction: (body?: any) => Promise<T>,
    immediate = true,
) => {
    const [status, setStatus] = useState<Status>('idle');
    const [value, setValue] = useState<T | null>(null);
    const [error, setError] = useState<E | null>(null);

    const execute = useCallback(
        (body?: any) => {
            setStatus('pending');
            setValue(null);
            setError(null);

            return asyncFunction(body)
                .then((response: any) => {
                    setValue(response);
                    setStatus('success');
                })
                .catch((error: any) => {
                    setError(error);
                    setStatus('error');
                });
        },
        [asyncFunction],
    );

    useEffect(() => {
        if (immediate) {
            execute();
        }
    }, [execute, immediate]);

    return { execute, status, value, error };
};

export default useAsync;
