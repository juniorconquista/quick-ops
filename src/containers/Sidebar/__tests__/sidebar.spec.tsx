import React from 'react';
import { cleanup, render } from '@testing-library/react';

import { SidebarFaker } from '../../../mocks/sidebar.faker';
import * as useAsync from '../../../hooks/useAsync';
import ThemeProvider from '../../../utils/ThemeProviderTest';
import Sidebar from '../sidebar';
import 'jest-styled-components';

const setupFetchStub: any = (data: any) => (_url: string) =>
    new Promise((resolve) => {
        resolve({
            json: () =>
                Promise.resolve({
                    data,
                }),
        });
    });

const sidebarResponseFaker = new SidebarFaker();
const sidebarResponse = sidebarResponseFaker.generate();

const props = {
    isOpen: false,
};

describe('Sidebar container', () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });

    beforeEach(() => {
        // spyOn's
        const fakeData = { fake: 'data' };
        jest.spyOn(global, 'fetch').mockImplementation(
            setupFetchStub(fakeData),
        );

        const useAsyncSpy = jest.spyOn(useAsync, 'default');
        const mockUseAsyncSpyFn = {
            ...jest.requireActual('../../../hooks/useAsync'),
            execute: jest.fn(),
            value: sidebarResponse,
        };
        useAsyncSpy.mockImplementation(
            jest.fn((fn) => {
                fn();
                return mockUseAsyncSpyFn;
            }),
        );
    });

    it('tests basic snapshot', () => {
        const useAsyncSpy = jest.spyOn(useAsync, 'default');
        const mockUseAsyncSpyFn = {
            ...jest.requireActual('../../../hooks/useAsync'),
            execute: jest.fn(),
            value: {},
        };
        useAsyncSpy.mockImplementation(
            jest.fn((fn) => {
                fn();
                return mockUseAsyncSpyFn;
            }),
        );

        const container = render(
            <ThemeProvider>
                <Sidebar {...props} />
            </ThemeProvider>,
        );
        expect(container.asFragment()).toMatchSnapshot();
    });
});
