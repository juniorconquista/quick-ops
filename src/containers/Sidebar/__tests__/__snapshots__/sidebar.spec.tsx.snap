// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar container tests basic snapshot 1`] = `
<DocumentFragment>
  .c3 {
  color: #f8f9fa;
  font-size: 2rem;
  font-weight: normal;
  color: #009d43;
  font-size: 16px;
  margin-bottom: 8px;
}

.c4 {
  color: #f8f9fa;
  font-size: 1.5rem;
  color: #009d43;
  font-size: 20px;
  font-weight: bold;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 40px 0px;
  width: 100%;
}

.c1:focus {
  outline: none;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #edfff5;
  padding: 0px 32px;
  min-height: 86px;
  width: 80%;
  min-width: 250px;
  border: 1px solid;
  border-radius: 6px;
  border-color: #009d43;
}

.c2:focus {
  outline: none;
}

.c0 {
  height: 100%;
}

.c0 .content {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: calc(100vh - 48px);
  height: 100%;
  background-color: #25262a;
  -webkit-transition: right .25s ease;
  transition: right .25s ease;
  position: relative;
}

.c0 .content:focus {
  outline: none;
}

.c0 .mobile {
  top: 48px;
  right: -350px;
  position: fixed;
  width: 350px;
}

.c0 .mobile.open {
  right: 0;
}

<div
    class="c0"
  >
    <div
      class="content mobile"
    >
      <div
        class="c1"
        width="100%"
      >
        <div
          class="c2"
          width="80%"
        >
          <h1
            class="c3"
            color="#009d43"
            font-size="16px"
          >
            Status
          </h1>
          <h2
            class="c4"
            color="#009d43"
            font-size="20px"
            font-weight="bold"
          >
            Conclu√≠do
          </h2>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
